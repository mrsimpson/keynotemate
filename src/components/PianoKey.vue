<template>
  <div
    :class="{ 'black-key': isBlack, 'white-key': !isBlack, active: isPressed }"
    class="piano-key"
    :style="getKeyStyle"
    @mousedown="playNote"
    @mouseup="stopNote"
    @mouseleave="stopNote"
    @touchstart.prevent="playNote"
    @touchend.prevent="stopNote"
  >
    <div class="key-label">{{ displayName }}</div>
  </div>
</template>

<script setup lang="ts">
import { computed, inject, ref } from 'vue';
import type { AudioService } from '../services/audio';
import { getPianoKeyDisplayName } from '../types/piano';
import { usePianoKey } from '../composables/usePianoKey';

const props = defineProps<{
  name: string;
  isBlack: boolean;
  color?: string;
  pitch: string;
  octave: number;
}>();

const audioService = inject<AudioService>('audioService');
const isPressed = ref(false);

const { getKeyStyle } = usePianoKey(props);

const displayName = computed(() => {
  return getPianoKeyDisplayName({ name: props.name, isBlack: props.isBlack });
});

const playNote = () => {
  isPressed.value = true;
  audioService?.playNote(props.pitch, props.octave, 0.5); // Default to octave 4 and 0.5s duration
};

const stopNote = () => {
  isPressed.value = false;
};
</script>

<style scoped>
.piano-key {
  position: relative;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: flex-end;
  padding-bottom: 0.5rem;
  user-select: none;
  border-radius: 0 0 4px 4px;
  cursor: pointer;
  transition: transform 0.1s ease;
}

.white-key {
  background: white;
  height: 300px;
  width: 80px;
  z-index: 1;
  border: 1px solid #ccc;
  border-bottom-width: 4px;
}

.black-key {
  background: #333;
  height: 150px;
  width: 40px;
  z-index: 2;
  border: 1px solid #000;
  border-bottom-width: 4px;
}

.key-label {
  position: absolute;
  bottom: 5px;
  font-size: 12px;
  font-weight: bold;
  color: #000;
}

.black-key .key-label {
  color: #fff;
}

.active {
  transform: translateY(2px);
}

.white-key.active {
  background-color: #f0f0f0;
}

.black-key.active {
  background-color: #222;
}
</style>
