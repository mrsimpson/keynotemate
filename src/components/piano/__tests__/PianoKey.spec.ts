import { mount } from '@vue/test-utils'
import { describe, expect, it } from 'vitest'
import PianoKey from '@/components/piano/PianoKey.vue'

describe('PianoKey Component', () => {
  const baseProps = {
    name: 'C4',
    isBlack: false,
    pitch: 'C',
    octave: 4,
  }

  it('renders white key snapshot', () => {
    const wrapper = mount(PianoKey, { props: baseProps })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders black key snapshot', () => {
    const wrapper = mount(PianoKey, {
      props: { ...baseProps, isBlack: true },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
