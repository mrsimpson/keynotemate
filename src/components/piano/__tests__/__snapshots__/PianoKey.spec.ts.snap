// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PianoKey Component > renders black key snapshot 1`] = `
"<div data-v-a7454f0c="" class="black-key piano-key">
  <div data-v-a7454f0c="" class="key-label">C4</div>
</div>"
`;

exports[`PianoKey Component > renders white key snapshot 1`] = `
"<div data-v-a7454f0c="" class="white-key piano-key">
  <div data-v-a7454f0c="" class="key-label">C4</div>
</div>"
`;
